src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkBatchInsertRecord.java

```diff
@@ -0,0 +1,322 @@
+
+            case Types.DATE:
+            case Types.CHAR:
+            case Types.NCHAR:
+            case Types.VARCHAR:
+            case Types.NVARCHAR:
+            case Types.LONGVARCHAR:
+            case Types.LONGNVARCHAR:
+            case Types.CLOB:
+            default: {
+                // The string is copied as is.
+                return data;
+            }
+        }
+    }
+
+    private String removeSingleQuote(String s) {
+        int len = s.length();
+        return (s.charAt(0) == '\'' && s.charAt(len - 1) == '\'') ? s.substring(1, len - 1) : s;
+    }
+
+    @Override
+    public Object[] getRowData() throws SQLServerException {
+        Object[] data = new Object[columnMetadata.size()];
+        int valueIndex = 0;
+        String valueData;
+        Object rowData;
+        int columnListIndex = 0;
+        
+        // check if the size of the list of values = size of the list of columns (which is optional)
+        if (null != columnList && columnList.size() != valueList.size()) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_CSVDataSchemaMismatch"));
+            Object[] msgArgs = {};
+            throw new SQLServerException(form.format(msgArgs), SQLState.COL_NOT_FOUND, DriverError.NOT_SET, null);
+        }
+        
+        for (Entry<Integer, ColumnMetadata> pair : columnMetadata.entrySet()) {
+            int index = pair.getKey() - 1;
+
+            // To explain what each variable represents:
+            // columnMetadata = map containing the ENTIRE list of columns in the table.
+            // columnList = the *optional* list of columns the user can provide. For example, the (c1, c3) part of this query: INSERT into t1 (c1, c3) values (?, ?)
+            // valueList = the *mandatory* list of columns the user needs provide. This is the (?, ?) part of the previous query. The size of this valueList will always equal the number of
+            // the entire columns in the table IF columnList has NOT been provided. If columnList HAS been provided, then this valueList may be smaller than the list of all columns (which is columnMetadata).
+            
+            // case when the user has not provided the optional list of column names.
+            if (null == columnList || columnList.size() == 0) {
+                valueData = valueList.get(index);
+                // if the user has provided a wildcard for this column, fetch the set value from the batchParam.
+                if (valueData.equalsIgnoreCase("?")) {
+                    rowData = batchParam.get(batchParamIndex)[valueIndex++].getSetterValue();
+                }
+                else if (valueData.equalsIgnoreCase("null")) {
+                    rowData = null;
+                }
+                // if the user has provided a hardcoded value for this column, rowData is simply set to the hardcoded value.
+                else {
+                    rowData = removeSingleQuote(valueData);
```

Comment:
Please test `insert into t values ('''')`

Reply:
It works. I've added this scenario as a test case as well.